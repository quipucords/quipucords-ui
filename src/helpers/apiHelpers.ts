/**
 * Extracts and concatenates error messages from an object.
 *
 * @param { { [key: string]: unknown } | undefined } data - The data object that may contain error messages.
 * @returns {string} A string containing error messages, joined by newlines, or 'Unknown error' if none are found.
 */
const extractErrorMessage = (data: { [key: string]: unknown } | undefined) => {
  if (!data) {
    return 'Unknown error';
  }
  const errorValues = Object.values(data);
  const messages = errorValues.flatMap(value => (Array.isArray(value) ? value : [value])).filter(Boolean);
  return messages.join('\n');
};

const apiHelpers = { extractErrorMessage };

export { apiHelpers as default, apiHelpers };
