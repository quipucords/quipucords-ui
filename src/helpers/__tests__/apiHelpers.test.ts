import apiHelpers from '../apiHelpers';

describe('apiHelpers', () => {
  it('should handle error messages from an object', () => {
    expect({
      'concatenate error messages from an object': apiHelpers.extractErrorMessage({
        error1: 'Error message 1',
        error2: 'Error message 2'
      }),
      'handle arrays of error messages': apiHelpers.extractErrorMessage({
        errors: ['Error message 1', 'Error message 2']
      }),
      'handle nested arrays of error messages': apiHelpers.extractErrorMessage({
        errors: ['Error message 1', ['Error message 2', 'Error message 3']]
      }),
      'filter out falsy values': apiHelpers.extractErrorMessage({
        error1: 'Error message 1',
        error2: null,
        error3: undefined,
        error4: '',
        error5: 'Error message 2'
      }),
      'return "Unknown error" for undefined input': apiHelpers.extractErrorMessage(undefined),
      'handle objects with mixed types of values': apiHelpers.extractErrorMessage({
        error1: 'Error message 1',
        error2: ['Error message 2', false, 0, 'Error message 3'],
        error3: null,
        error4: { nestedError: 'Nested error message' }
      })
    }).toMatchSnapshot('extractErrorMessage');
  });
});
