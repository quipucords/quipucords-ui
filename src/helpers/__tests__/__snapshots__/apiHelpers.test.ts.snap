// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`apiHelpers should handle error messages from an object: extractErrorMessage 1`] = `
{
  "concatenate error messages from an object": "Error message 1
Error message 2",
  "filter out falsy values": "Error message 1
Error message 2",
  "handle arrays of error messages": "Error message 1
Error message 2",
  "handle nested arrays of error messages": "Error message 1
Error message 2,Error message 3",
  "handle objects with mixed types of values": "Error message 1
Error message 2
Error message 3
[object Object]",
  "return "Unknown error" for undefined input": "Unknown error",
}
`;
